addons/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/discovery/KM200GatewayDiscoveryService.java

```diff
@@ -0,0 +1,166 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link KM200GatewayDiscoveryService} class discovers things through a gateway
+ *
+ * @author Markus Eckhardt
+ *
+ */
+public class KM200GatewayDiscoveryService extends AbstractDiscoveryService implements KM200GatewayStatusListener {
+
+    private Logger logger = LoggerFactory.getLogger(KM200GatewayDiscoveryService.class);
+
+    private static int TIMEOUT = 120;
+
+    KM200GatewayHandler gateway;
+
+    public KM200GatewayDiscoveryService(KM200GatewayHandler gateway) {
+        super(KM200GatewayHandler.SUPPORTED_THING_TYPES_UIDS, TIMEOUT, true);
+        logger.debug("KM200GatewayDiscoveryService {}", gateway);
+        this.gateway = gateway;
+        this.gateway.addGatewayStatusListener(this);
+    }
+
+    @Override
+    protected void startScan() {
+        discoverDevices();
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        discoverDevices();
+    };
+
+    @Override
+    public void gatewayStatusChanged(ThingStatus status) {
+        if (status.equals(ThingStatus.ONLINE)) {
+            discoverDevices();
+        }
+    }
+
+    @Override
+    protected void deactivate() {
+        super.deactivate();
+        gateway.removeHubStatusListener(this);
+    }
+
+    /**
+     * Discovers devices connected to a hub
+     */
+    private void discoverDevices() {
+        if (!gateway.device.getInited()) {
+            logger.info("Gateway not configured, scanning postponed.");
+            return;
+        }
+        ThingUID thingUID = null;
+        ThingUID bridgeUID = gateway.getThing().getUID();
+        logger.debug("Checking bridgeUID: {} ", bridgeUID.getAsString());
+
+        for (KM200ThingTypes tType : KM200ThingTypes.values()) {
+            String root = tType.getRootPath();
+            if (root.isEmpty()) {
+                continue;
+            }
+            String checkService = tType.getActiveCheckSubPath();
+            logger.debug("root: {}", root);
+            if (gateway.device.containsService(root)) {
+                Boolean enumOnly = true;
+                KM200CommObject object = gateway.device.getServiceObject(root);
+                Set<String> keys = object.serviceTreeMap.keySet();
+                /* Check whether all sub services are refEnum */
+                for (String key : keys) {
+                    if (!object.serviceTreeMap.get(key).getServiceType().equals("refEnum")) {
+                        enumOnly = false;
+                        break;
+                    }
+                }
+                /* If there are refEnum only, then create for every one an own thing */
+                if (enumOnly) {
+                    for (String key : keys) {
+                        /* Check whether this part of heating system is inactive. If its then ignore it */
+                        if (checkService != null) {
+                            if (gateway.device.containsService(root + "/" + key + "/" + checkService)) {
```

Comment:
Perhaps you could assign `gateway.device.containsService(root + "/" + key + "/" + checkService)` to the `bool` variable and use it both in this condition and in :110 with `.getValue()`